days/src/day7/Day7.kt

```diff
@@ -0,0 +1,66 @@
+package day7
+
+import Day
+import java.io.File
+
+class Day7(private val bagRules: Map<String, List<Pair<Int, String>>>) : Day {
+    private lateinit var containmentGraph: Map<String, List<String>>
+
+    init {
+        val graph: MutableMap<String, MutableList<String>> = mutableMapOf()
+        bagRules.forEach { (containingBag, contained) ->
+            contained.forEach { (amount, bag) ->
+                val newEdgeList = graph.getOrDefault(bag, mutableListOf())
+                newEdgeList.add(containingBag)
+                graph[bag] = newEdgeList
+            }
+        }
+        containmentGraph = graph.toMap().mapValues { (_, values) -> values.toList() }
+    }
+
+    private fun containmentBfs(start: String): Map<String, List<String>> {
+        val outerBags = mutableMapOf<String, List<String>>()
+        data class BfsNode(val node: String, val history: List<String>) {
+            val distance: Int
+                get() = history.size
+        }
+        val wavefront = mutableListOf<BfsNode>()
+        wavefront.add(BfsNode(start, listOf()))
+        while (wavefront.isNotEmpty()) {
+            val current = wavefront.removeAt(0)
+            val containingBags = containmentGraph[current.node]
+            outerBags[current.node] = current.history
+            if (containingBags != null && containingBags.isNotEmpty()) {
+                containingBags.forEach { bag ->
+                    wavefront.add(BfsNode(bag, listOf(current.node) + current.history))
+                }
+            }
+        }
+        outerBags.remove(start)
+        return outerBags.toMap()
+    }
+
+    private fun containsDfs(start: String): Int = bagRules[start]!!.map {
+        it.first * (containsDfs(it.second) + 1) }.sum()
```

Comment:
This is pretty sweet, is the `bagRules[start]!!` the base case?

Reply:
The !! is actually the unsafe null operator. It happens because by construction all bags should be in there. It should be using `?.map{ ...}.?sum() ?: 0` probably to be safe. 
The base case is by construction. Innermost bags have empty containment lists so the `.map{}` does nothing and `.sum()` return 0